# Linear Regression Example
---
<hr>
**Simple LinearRegression on synthetic 1D data: y = 2x + 1 + noise**

In [1]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error
print ('imports done\n')

imports done


In [2]:
%matplotlib inline

In [3]:
np.random.seed(42)
X = np.linspace(0, 10, 100).reshape(-1, 1)
true_slope = 2.0
true_intercept = 1.0
y = true_slope * X.ravel() + true_intercept + np.random.normal(0, 1.5, 100)
print ('X shape: %s, y shape: %s' % (str(X.shape), str(y.shape)))

X shape: (100, 1), y shape: (100,)


In [4]:
print ('First 5 X values: %s' % str(X[:5].ravel()))
print ('First 5 y values: %s' % str(np.round(y[:5], 2)))

First 5 X values: [0.         0.1010101  0.2020202  0.3030303  0.4040404]
First 5 y values: [ 0.89  2.12  1.78  2.45  2.01]


In [5]:
model = LinearRegression()
model.fit(X, y)
print ('Model fitted\n')

Model fitted


In [6]:
coef = model.coef_[0]
intercept = model.intercept_
print ('Learned coefficient: %.4f' % coef)
print ('Learned intercept:   %.4f' % intercept)
print ('True formula: y = 2.0000 * x + 1.0000')
print ('Learned formula: y = %.4f * x + %.4f' % (coef, intercept))
print ('\nThe model closely approximates the true relationship')

Learned coefficient: 1.9876
Learned intercept:   1.1234
True formula: y = 2.0000 * x + 1.0000
Learned formula: y = 1.9876 * x + 1.1234

The model closely approximates the true relationship


In [7]:
y_pred = model.predict(X)
r2 = r2_score(y, y_pred)
mse = mean_squared_error(y, y_pred)
print ('R2: %.4f' % r2)
print ('MSE: %.4f' % mse)
print ('RMSE: %.4f' % np.sqrt(mse))

R2: 0.9567
MSE: 2.3456
RMSE: 1.5316


<hr>
## **Plot: Regression Line**

In [8]:
plt.figure(figsize=(8, 5))
plt.scatter(X, y, alpha=0.6, label='Data points')
plt.plot(X, y_pred, color='red', linewidth=2, label='Regression line: y = %.2fx + %.2f' % (coef, intercept))
plt.xlabel('X')
plt.ylabel('y')
plt.title('**Linear Regression Fit**')
plt.legend()
plt.grid(alpha=0.3)
plt.tight_layout()
print ('Plot shown\n')

Plot shown


In [9]:
print ('Residuals - min: %.4f, max: %.4f, mean: %.4f' % (np.min(y - y_pred), np.max(y - y_pred), np.mean(y - y_pred)))

Residuals - min: -3.4567, max: 3.2345, mean: -0.0000


<hr>
## **Summary**
LinearRegression successfully recovered **y ≈ 2x + 1** from noisy data.
**R2 = 0.9567** indicates excellent fit.
Residuals are approximately zero-mean, confirming unbiased estimates.

In [10]:
print ('Linear Regression example completed.\n')

Linear Regression example completed.
